dspace-api/src/main/java/org/dspace/google/GoogleAsyncEventListener.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.log4j.Logger;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.core.Constants;
+import org.dspace.core.Context;
+import org.dspace.services.ConfigurationService;
+import org.dspace.services.model.Event;
+import org.dspace.usage.AbstractUsageEventListener;
+import org.dspace.usage.UsageEvent;
+import org.springframework.beans.factory.annotation.Autowired;
+
+
+/**
+ * Notifies Google Analytics of Bitstream VIEW events. These events are stored in memory and then
+ * asynchronously processed by a single seperate thread.
+ *
+ * @author April Herron
+ */
+public class GoogleAsyncEventListener extends AbstractUsageEventListener {
+
+    private static final int MAX_TIME_SINCE_EVENT = 14400000;
+    private static final String ANALYTICS_BATCH_ENDPOINT = "https://www.google-analytics.com/batch";
+    private static Logger log = Logger.getLogger(GoogleAsyncEventListener.class);
+    private static String analyticsKey;
+    private static CloseableHttpClient httpclient;
+    private static Buffer buffer;
+    private static ExecutorService executor;
+    private static Future future;
+    private static boolean destroyed = false;
+
+    @Autowired(required = true)
+    ConfigurationService configurationService;
+
+    @PostConstruct
+    public void init() {
+        analyticsKey = configurationService.getProperty("google.analytics.key");
+        if (StringUtils.isNotEmpty(analyticsKey)) {
+            int analyticsBufferlimit = configurationService.getIntProperty("google.analytics.buffer.limit", 256);
+            buffer = BufferUtils.synchronizedBuffer(new CircularFifoBuffer(analyticsBufferlimit));
+            httpclient = HttpClients.createDefault();
+            executor = Executors.newSingleThreadExecutor();
+            future = executor.submit(new GoogleAnalyticsTask());
+        }
+    }
+
+    @Override
+    public void receiveEvent(Event event) {
+        if ((event instanceof UsageEvent)) {
+            if (StringUtils.isNotEmpty(analyticsKey)) {
+                UsageEvent ue = (UsageEvent) event;
+                log.debug("Usage event received " + event.getName());
+                try {
+                    if (ue.getAction() == UsageEvent.Action.VIEW &&
+                            ue.getObject().getType() == Constants.BITSTREAM) {
+
+                        // Client Id, should uniquely identify the user or device. If we have a session id for the user
+                        // then lets use it, else generate a UUID.
+                        String cid;
+                        if (ue.getRequest().getSession(false) != null) {
+                            cid = ue.getRequest().getSession().getId();
```

Comment:
For DSpace 7, we don't really have an HTTPSession.  Instead, for the Client ID, it might be better to consider using the `X-CORRELATION-ID` header (if it exists), as that's also what we use in the logs as a replacement for a session id.  See #3303